proto/p4/p4types.proto

```diff
@@ -0,0 +1,207 @@
+    // multiple values. Unlike a struct type, tuples have no named fields."
+    P4TupleTypeSpec tuple = 3;
+    P4NamedType struct = 4;
+    P4NamedType header = 5;
+    P4NamedType header_union = 6;
+    P4HeaderStackTypeSpec header_stack = 7;
+    P4HeaderUnionStackTypeSpec header_union_stack = 8;
+    P4NamedType enum = 9;
+    P4ErrorType error = 10;
+  }
+}
+
+message P4NamedType {
+  string name = 1;
+}
+
+// Empty message as no type information needed, just used as a placeholder in
+// the oneof to identify boolean types.
+message P4BoolType { }
+
+message P4ErrorType { }
+
+message P4BitstringLikeTypeSpec {
+  oneof type_spec {
+    P4BitTypeSpec bit = 1;  // bit<W>
+    P4IntTypeSpec int = 2;  // int<W>
+    P4VarbitTypeSpec varbit = 3;  // varbit<W>
+  }
+  // Useful to identify well-know types, such as IP address or Ethernet MAC
+  // address.
+  repeated string annotations = 4;
+}
+
+message P4BitTypeSpec {
+  int32 bitwidth = 1;
+}
+
+message P4IntTypeSpec {
+  int32 bitwidth = 1;
+}
+
+message P4VarbitTypeSpec {
+  int32 max_bitwidth = 1;
+}
+
+message P4TupleTypeSpec {
+  repeated P4DataTypeSpec members = 1;
+}
+
+message P4StructTypeSpec {
+  message Member {
+    string name = 1;
+    P4DataTypeSpec type_spec = 2;
+  }
+  repeated Member members = 1;
+  repeated string annotations = 2;
+}
+
+message P4HeaderTypeSpec {
+  message Member {
+    string name = 1;
+    P4BitstringLikeTypeSpec type_spec = 2;
+  }
+  repeated Member members = 1;
+  repeated string annotations = 2;
+}
+
+message P4HeaderUnionTypeSpec {
+  message Member {
+    string name = 1;
+    P4NamedType header = 2;
+  }
+  repeated Member members = 1;
+  repeated string annotations = 2;
+}
+
+message P4HeaderStackTypeSpec {
+  P4NamedType header = 1;
+  int32 size = 2;
+}
+
+message P4HeaderUnionStackTypeSpec {
+  P4NamedType header_union = 1;
+  int32 size = 2;
+}
+
+// As of today, P4_16 enum members have no integer value assigned to
+// them. Future versions of this protobuf interface may use integer values for
+// enums in P4Data, which means the repeated field would be replaced by a map in
+// this message (mapping from member names to integer values). This is dependent
+// on future versions of the P4 language.
+message P4EnumTypeSpec {
+  repeated string members = 1;
+  repeated string annotations = 2;
+}
+
+// Similar to an enum, but there is always one and only one instance per P4
+// program.
+message P4ErrorTypeSpec {
+  repeated string members = 1;
+}
+
+// End of P4 type specs --------------------------------------------------------
+
+message P4Data {
+  oneof data {
+    bytes bitstring = 1;  // for bit<W>, int<W>, varbit<W>
+    bool bool = 2;
+    P4StructLike tuple = 3;
+    P4StructLike struct = 4;
+    P4Header header = 5;
+    P4HeaderUnion header_union = 6;
+    P4HeaderStack header_stack = 7;
+    P4HeaderUnionStack header_union_stack = 8;
+    // Could be replaced with integer values in future versions.
```

Comment:
I suppose that in the future, some programmers may choose to assign integer values to some subset of enums, while others may choose to retain named enums, so wouldn't this need to allow for both?

Reply:
I believe that before the programmer is allowed to choose integers, the language will need to describe how integer values are assigned **by default**, i.e. with programmer intervention. Which means that even when the programmer doesn't provide a specific value, the compiler assigns one. If this change happens in the P4 language, we can safely deprecate the `string` field and add `int32` or `int64` fields here, with a mapping in P4Info (name -> int), since all enum members will always have a integer value.

Reply:
I can imagine the P4_16 language spec allowing some enum's to have _no_ definition of integer values for enum members, in a case where they never need to communicate that enum value into or out of the P4 program to other places (whether that other place is the control plane, or another data plane device).

Whether there is enough of an advantage for the language to require the programmer to specify integer values for all enum members for those enum types that need them, and leave out integer value specifications by the programmer for other enum types that do not need them, I can raise that question in the LDWG when it comes up.  I don't have a strong opinion either way on that detail.

I have a mild opinion that having some enums with integer values specified, and other enums without the integer values specified, adds extra error checks and 2 cases to think about in lots of places involving enum's, instead of only 1 case.